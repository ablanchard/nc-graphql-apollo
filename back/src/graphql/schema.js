import { makeExecutableSchema } from 'graphql-tools'
import types from './type'
import { resolvers as queryResolvers, queries } from './query'
import { resolvers as mutationResolvers, mutations } from './mutation'
import { resolvers as subscriptionResolvers, subscriptions} from './subscription'
import { merge } from 'lodash'

const typeDefs = [queries, mutations, subscriptions, types].join()

const resolvers = merge(queryResolvers, mutationResolvers, subscriptionResolvers)

export default makeExecutableSchema({
    typeDefs,
    resolvers
})