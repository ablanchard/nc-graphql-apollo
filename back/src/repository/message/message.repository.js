let id = 0
let messages = []

export const getMessages = function() {
    return messages
}

export const addMessage = function(message) {
    messages.push(message)
    return messages[messages.length - 1]
}

export const resetRepository = function() {
    messages = []
}

export const getNextId = function() {
    return id++
}