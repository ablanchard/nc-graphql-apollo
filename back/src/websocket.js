import { execute, subscribe } from 'graphql'
import { SubscriptionServer } from 'subscriptions-transport-ws'
import { subscriptionsEndPoint } from './router'
import schema from './graphql/schema'

export const initWebsocket = function(websockerServer) {
    new SubscriptionServer({
        execute,
        subscribe,
        schema
    }, {
        server: websockerServer,
        path: `/${subscriptionsEndPoint}`
    })    
}