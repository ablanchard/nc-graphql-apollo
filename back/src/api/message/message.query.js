import { getMessages } from '../../service/message/message.service'

const MessageType = `
    type Message {
        id: ID!
        date: String!
        sender: Sender!
        content: String!
        localisation: String!
        status: String!
    }
`

const SenderType = `
    type Sender {
        pseudo: String!
        firstName: String!
        lastName: String!
    }
`

export const queries = `
    getMessages : [Message!]!
`

export const types = `
    ${MessageType}
    ${SenderType}
`

export const resolvers = {
    getMessages: () => {
        return getMessages()
    }
}