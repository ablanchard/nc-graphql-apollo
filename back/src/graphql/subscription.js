import { merge } from 'lodash'
import { subscriptions as messageSubscriptions, resolvers as messageResolvers } from '../schema/message/message.subscription'

export const subscriptions = `
    type Subscription {
        ${messageSubscriptions}
    }
`
export const resolvers = {
    Subscription: merge(
        messageResolvers,
    )
}