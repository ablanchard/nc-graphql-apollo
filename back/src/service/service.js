import { init as initMessageSubscriber } from './message/message.subscriber'

export const initServices = async function() {
    await initMessageSubscriber()
}