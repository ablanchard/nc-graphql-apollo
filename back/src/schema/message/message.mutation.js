import { addMessage } from '../../service/message/message.service'

const MessageInput = `
    input MessageInput {
        content: String!
        localisation: String!
        status: String!
        sender: SenderInput!
    }
`

const SenderInput = `
    input SenderInput {
        pseudo: String!
        firstName: String!
        lastName: String!
    }
`

export const mutations = `
    saveMessage(message: MessageInput): Message
`

export const types = `
    ${MessageInput}
    ${SenderInput}
`

export const resolvers = {
    saveMessage: (_, {message}) => {
        return addMessage(message)
    }
}