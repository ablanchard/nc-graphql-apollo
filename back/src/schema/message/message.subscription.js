import { pubsub } from '../../service/websocket/websocket.service'

export const types = `
`

export const subscriptions = `
    subscribeMessages: Message
`

export const resolvers = {
    subscribeMessages: {
        resolve: (payload) => payload,
        subscribe: () => pubsub.asyncIterator(['subscribeMessages'])
    }
}