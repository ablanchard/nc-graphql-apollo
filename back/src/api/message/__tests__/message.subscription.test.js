jest.mock('../../../service/websocket/websocket.service', () => {
    return {
        pubsub: {
            asyncIterator: jest.fn()
        }
    }
})
import { pubsub } from '../../../service/websocket/websocket.service'
import { types, subscriptions} from '../message.subscription'
import schema from '../../../graphql/schema'


describe('Schema of Contributor (Subscription)', () => {
    it('should be like snapshot (type)', () => {
        expect(types).toMatchSnapshot()
    })
    it('should be like snapshot (subscriptions)', () => {
        expect(subscriptions).toMatchSnapshot()
    })
})

describe('subscribeMessages function', () => {
    describe('resolve property', () => {
        let result
        beforeAll(() => {
            const resolver = schema.getSubscriptionType().getFields().subscribeMessages.resolve
            result = resolver('FAKE_PAYLOAD')
        })
        it('should have a resolve which return payload', () => {
            expect(result).toBe('FAKE_PAYLOAD')
        })
    })
    describe('subscribe property', () => {
        let result
        beforeAll(() => {
            const resolver = schema.getSubscriptionType().getFields().subscribeMessages.subscribe
            pubsub.asyncIterator.mockReturnValue('FAKE_RESULT_FROM_ASYNCITERATOR')
            result = resolver()
        })
        it('should have a function which call pub.asyncIterator with contributor event types as parameter', () => {
            expect(pubsub.asyncIterator).toHaveBeenCalledWith(['subscribeMessages'])
        })
        it('should return result from asyncIterator', () => {
            expect(result).toBe('FAKE_RESULT_FROM_ASYNCITERATOR')
        })
    })
})