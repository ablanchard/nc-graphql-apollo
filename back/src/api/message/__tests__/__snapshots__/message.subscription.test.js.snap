// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema of Contributor (Subscription) should be like snapshot (subscriptions) 1`] = `
"
    subscribeMessages: Message
"
`;

exports[`Schema of Contributor (Subscription) should be like snapshot (type) 1`] = `
"
"
`;
